### ISSUE TITLE
Infinite loop when using fetcher.load in latest Remix releases

### ISSUE THREAD

**Issue by BrycePearce** [remix-run/remix#6072]
### What version of Remix are you using?

1.14.0 (also on 1.15)

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

**1.)** Create a resource-route (or any loader) that returns anything. In this case I'm returning `json({hello: 'world'})`
**2.)** In any route, create a state variable to track to see if it has been invoked. For example

```js
  let fetcher = useFetcher();
  const [shouldCallResourceRoute, setShouldCallResourceRoute] = useState(true);

  useEffect(() => {
    if (shouldCallResourceRoute) {
      setShouldCallResourceRoute(false); // this should cause this fetcher to never load again. But it infinite loops
      fetcher.load("/resource-routes/resource");
    }
  }, [fetcher, shouldCallResourceRoute]);
````
**3.)** Note the infinite loop / crash that happens
![image](https://user-images.githubusercontent.com/16729071/232183037-fd12da86-7379-404d-8138-708bad6f8085.png)

### **-Note-**

 I am upgrading my Remix from an earlier version where my existing code worked fine. This bug only appeared after upgrading and I was able to replicate in a code sandbox here:

https://codesandbox.io/p/sandbox/hopeful-night-z2lsum?file=%2Fapp%2Froutes%2Findex.tsx&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A7%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A7%7D%5D

### Expected Behavior

The fetcher should only load once

### Actual Behavior

An infinite loop occurs, crashing the page.

**Comment by chuanhhoang** [comment]
I face this problem too. Any update on this?

**Comment by BrycePearce** [comment]
> I face this problem too. Any update on this?

I believe this is an unlisted bug/change that only affects the latest version. However there is a work around provided by brocktho in the discord chat

```js
  useEffect(() => {
    if (shouldFetch && fetcher.state === 'idle') {
      fetcher.load('/fetchRoute');
      setShouldFetch(false);
    }
  }, [fetcher, shouldFetch]);
  ```
  https://stackblitz.com/edit/remix-run-remix-ahjkhw?file=app%2Froutes%2F_index.tsx,app%2Fentry.server.tsx
  
  ----

However as an unlisted change that breaks previous code, I still believe this is a bug

**Comment by pkasarda** [comment]
probably dup of https://github.com/remix-run/remix/issues/6014

**Comment by vaynevayne** [comment]
this can work ,when remove fetcher from dependencies, so i think fetcher is not stable,

<img width="1249" alt="image" src="https://user-images.githubusercontent.com/124666577/232742151-5acc6bf0-740c-4c68-8761-ee3c14710290.png">

but  look like remix's  `useFetcher` has not problem
https://github.com/remix-run/remix/blob/main/packages/remix-react/components.tsx#L1459
```jsx
import {
  useFetcher as useFetcherRR,
} from "react-router-dom";

export function useFetcher<TData = any>(): FetcherWithComponents<
  SerializeFrom<TData>
> {
  let fetcherRR = useFetcherRR();

  return React.useMemo(() => {
    
    return fetcherWithComponents;
  }, [fetcherRR]);
}
```

next  find `react-router-dom`

```jsx
export function useFetcher<TData = any>(): FetcherWithComponents<TData> {

 let fetcher = router.getFetcher<TData>(fetcherKey); //  bug is here , This is variable

 let fetcherWithComponents = React.useMemo(
    () => ({
      Form,
      submit,
      load,
      ...fetcher,
    }),
    [fetcher, Form, submit, load]
  );
return  fetcherWithComponents
}

// This is a factory function
 getFetcher() {
      return IDLE_FETCHER;
    },

```

when `shouldCallResourceRoute` from `true` to `false` ->  reRender, exec useFetcher, invoke `router.getFetcher`,
-> fetcher changed  -> rerender -> ... infinity loop

**Comment by kiliman** [comment]
Yes, I believe that you should always check `fetcher.state` and `fetcher.data` and not rely on a stable fetcher instance.

**Comment by brophdawg11** [comment]
This turns out to be due to the usage of `useSyncExternalStore` inside of react router and explains why this was not an issue prior to `1.10.0`.  The tl;dr; is in https://github.com/remix-run/react-router/pull/10377, but this should be fixed once that gets merged and released.

**Comment by brophdawg11** [comment]
This is fixed in `1.16.0-pre.1`.  I was having trouble getting the codesandbox above to actually install the new version but I recreated it locally and upgrading to `1.16.0-pre.1` fixed the issue.  We're hoping to get the 1.16.0 stable release out in the next few days.
